File: fsharp/BronKerbosch/FortifiedCounter.fs
namespace BronKerbosch

open System.Diagnostics

// Counts the coming and going of elements and, in debug builds only, checks their identity.
type FortifiedCounter<'T when 'T: equality> =
    { mutable count: int
      mutable tracker: DebugOnlyTracker<'T> }

    static member init() : FortifiedCounter<'T> =
        { count = 0
          tracker = DebugOnlyTracker.init () }

    member this.Contains(element: 'T) = this.tracker.Contains(element)

    static member Add(this: byref<FortifiedCounter<'T>>, element: 'T) : Unit =
        Debug.Assert(this.count = this.tracker.Count)
        Debug.Assert(this.Contains element |> not)
        this.count <- this.count + 1
        DebugOnlyTracker.Add(&this.tracker, element)
        Debug.Assert(this.count = this.tracker.Count)

    static member Remove(this: byref<FortifiedCounter<'T>>, element: 'T) =
        Debug.Assert(this.count = this.tracker.Count)
        Debug.Assert(this.Contains element)
        this.count <- this.count - 1
        DebugOnlyTracker.Remove(&this.tracker, element)
        Debug.Assert(this.count = this.tracker.Count)

File: csharp/BronKerbosch/FortifiedCounter.cs
using System.Diagnostics;

namespace BronKerbosch
{
    // Counts the coming and going of elements and, in debug builds only, checks their identity.
    public sealed class FortifiedCounter<T>
    {
        public int Count { get; private set; }

        private readonly DebugOnlyTracker<T> Tracker = new();

        public bool Contains(T element) => Tracker.Contains(element);

        public void Add(T element)
        {
            Count += 1;
            Tracker.Add(element);
            Debug.Assert(Count == Tracker.Count);
        }

        public void Remove(T element)
        {
            Count -= 1;
            Trace.Assert(Count >= 0);
            Tracker.Remove(element);
            Debug.Assert(Count == Tracker.Count);
        }
    }
}
